executor/src/witgen/machines/write_once_memory.rs

```diff
@@ -0,0 +1,221 @@
+
+        // Build a Vec<PolyID> for the key and value polynomials
+        let (key_polys, value_polys): (Vec<_>, Vec<_>) = rhs_polys
+            .into_iter()
+            .partition(|p| p.poly_id.ptype == PolynomialType::Constant);
+        let key_polys = key_polys
+            .into_iter()
+            .map(|p| {
+                assert!(!p.next);
+                p.poly_id
+            })
+            .collect::<Vec<_>>();
+        let value_polys = value_polys
+            .into_iter()
+            .map(|p| {
+                assert!(!p.next);
+                p.poly_id
+            })
+            .collect::<Vec<_>>();
+
+        let mut key_to_index = BTreeMap::new();
+        for row in 0..fixed_data.degree {
+            let key = key_polys
+                .iter()
+                .map(|k| fixed_data.fixed_cols[k].values[row as usize])
+                .collect::<Vec<_>>();
+            if key_to_index.insert(key, row).is_some() {
+                // Duplicate keys, can't be a write-once memory
+                return None;
+            }
+        }
+
+        Some(Self {
+            fixed_data,
+            rhs,
+            value_polys,
+            key_to_index,
+            data: BTreeMap::new(),
+        })
+    }
+
+    fn process_plookup_internal(
+        &mut self,
+        left: &[AffineExpression<&'a AlgebraicReference, T>],
+        right: &'a SelectedExpressions<Expression<T>>,
+    ) -> EvalResult<'a, T> {
+        let (key_expressions, value_expressions): (Vec<_>, Vec<_>) = left
+            .iter()
+            .zip(right.expressions.iter())
+            .partition(|(_, r)| {
+                try_to_simple_poly(r).unwrap().poly_id.ptype == PolynomialType::Constant
+            });
+        let key = key_expressions
+            .into_iter()
+            .map(|(k, _)| k.constant_value())
+            .collect::<Option<Vec<_>>>();
+        let value_expressions = value_expressions
+            .into_iter()
+            .map(|(v, _)| v)
+            .collect::<Vec<_>>();
+        let value = value_expressions
+            .iter()
+            .map(|v| v.constant_value())
+            .collect::<Option<Vec<_>>>();
+
+        log::trace!("Key: {:?}", key);
+        log::trace!("Value: {:?}", value);
+
+        if key.is_none() {
+            return Ok(EvalValue::incomplete(
+                IncompleteCause::NonConstantRequiredArgument("key"),
+            ));
+        }
+        let key = key.unwrap();
+        let index = self.key_to_index.get(&key).cloned();
+
+        if index.is_none() {
+            return Err(format!("Key {:?} not found in write-once memory", key).into());
+        }
+        let index = index.unwrap();
+
+        // If there is an externally provided memory value, use it
+        let external_witness_value = self
+            .value_polys
+            .iter()
+            .map(|p| self.fixed_data.external_witness(index, p))
+            .collect::<Option<Vec<_>>>();
```

Comment:
Shouldn't this be a vector of options? Can it be that we have provided values for some but not all columns?

Reply:
Taking this further, could we have an instruction that writes to one column and reads from another on the same row in the write once machine?

Reply:
> Shouldn't this be a vector of options? Can it be that we have provided values for some but not all columns?

True, I so far assumed that either the entire value is read / written or nothing happens. I generalized this in: 3ef9ed0fb2656ff1fa5f3c2d1ef25033821896b9

> Taking this further, could we have a single instruction that writes to one column and reads from another on the same row at the same time in the write once machine?

I think that should be fine, at least after the last commit, since the read and write lookup is identical!